src/radiation_system.hpp

```diff
@@ -132,19 +150,76 @@ template <typename problem_t> class RadSystem : public HyperbolicSystem<problem_
+	AMREX_GPU_HOST_DEVICE static auto ComputeFluxMeanOpacity(double rho, double Tgas) -> quokka::valarray<double, nGroups_>;
+	AMREX_GPU_HOST_DEVICE static auto ComputePlanckOpacityTempDerivative(double rho, double Tgas) -> quokka::valarray<double, nGroups_>;
 	AMREX_GPU_HOST_DEVICE static auto ComputeEintFromEgas(double density, double X1GasMom, double X2GasMom, double X3GasMom, double Etot) -> double;
 	AMREX_GPU_HOST_DEVICE static auto ComputeEgasFromEint(double density, double X1GasMom, double X2GasMom, double X3GasMom, double Eint) -> double;
 
+	// CCH: linear equation solver for matrix with non-zeros at the first row, first column, and diagonal only
+	AMREX_GPU_HOST_DEVICE static void SolveLinearEqs(double a00, const quokka::valarray<double, nGroups_> &a0i,
+							 const quokka::valarray<double, nGroups_> &ai0, const quokka::valarray<double, nGroups_> &aii,
+							 const double &y0, const quokka::valarray<double, nGroups_> &yi, double &x0,
+							 quokka::valarray<double, nGroups_> &xi);
+
+	AMREX_GPU_HOST_DEVICE static auto ComputePlanckEnergyFractions(amrex::GpuArray<double, nGroups_ + 1> const &boundaries, amrex::Real temperature)
+	    -> quokka::valarray<amrex::Real, nGroups_>;
+
 	template <FluxDir DIR>
 	AMREX_GPU_DEVICE static auto ComputeCellOpticalDepth(const quokka::Array4View<const amrex::Real, DIR> &consVar,
-							     amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx, int i, int j, int k) -> double;
+							     amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx, int i, int j, int k)
+	    -> quokka::valarray<double, nGroups_>;
 
 	AMREX_GPU_DEVICE static auto isStateValid(std::array<amrex::Real, nvarHyperbolic_> &cons) -> bool;
 };
 
+// CCH: compute radiation energy fractions for each photon group from a Planck function, given nGroups, radBoundaries, and temperature
+template <typename problem_t>
+AMREX_GPU_HOST_DEVICE auto RadSystem<problem_t>::ComputePlanckEnergyFractions(amrex::GpuArray<double, nGroups_ + 1> const &boundaries, amrex::Real temperature)
+    -> quokka::valarray<amrex::Real, nGroups_>
+{
+	quokka::valarray<amrex::Real, nGroups_> radEnergyFractions{};
+	if constexpr (nGroups_ == 1) {
+		// CCH: for future reference, if we want to use a single group as a delta function, say at 13.6 eV, rethink this.
+		radEnergyFractions[0] = 1.0;
+		return radEnergyFractions;
+	} else {
+		amrex::Real const energy_unit_over_kT = RadSystem_Traits<problem_t>::energy_unit / (boltzmann_constant_ * temperature);
+		amrex::Real previous = integrate_planck_from_0_to_x(boundaries[0] * energy_unit_over_kT);
+		for (int g = 0; g < nGroups_; ++g) {
+			amrex::Real y = integrate_planck_from_0_to_x(boundaries[g + 1] * energy_unit_over_kT);
+			radEnergyFractions[g] = y - previous;
+			previous = y;
+		}
+		// A hack to avoid zeros in radEnergyFractions
+		// TODO(CCH): find a better way to do this
+		for (int g = 0; g < nGroups_; ++g) {
+			if (radEnergyFractions[g] < 1e-14) {
+				radEnergyFractions[g] = 1e-14;
```

Comment:
Can you change this to be a variable instead of a hard-coded value of `1e-14`? A default value for this variable of 1e-14 seems fine to me, but it should not be hard-coded if we can avoid it.

Reply:
Done.